File: cpp/hot100/50.二叉树中的最大路径和.cpp
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        auto dfs = [&](this auto&& dfs, TreeNode* node) -> int{
            if(node == nullptr){
                return 0;
            }
            int left = dfs(node->left);
            int right = dfs(node->right);
            ans = max(ans, left + right + node->val);
            return max(max(left, right) + node->val, 0);
        };
        dfs(root);
        return ans;
    }
};

File: python/hot100/50.二叉树中的最大路径和.py
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.ans = -inf
        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            self.ans = max(left + right + node.val, self.ans)
            return max(max(left, right) + node.val, 0)
        dfs(root)
        return self.ans